src/GafferScene/LightToCamera.cpp

```diff
@@ -335,3 +385,139 @@ IECore::ConstCompoundObjectPtr LightToCamera::computeProcessedAttributes( const
+}
+
+
+IECore::ConstInternedStringVectorDataPtr LightToCamera::computeSetNames( const Gaffer::Context *context, const ScenePlug *parent ) const
+{
+	ConstInternedStringVectorDataPtr inputSetNamesData = inPlug()->setNamesPlug()->getValue();
+	const std::vector<InternedString> &inNames = inputSetNamesData->readable();
+
+	if( std::find( inNames.begin(), inNames.end(), g_camerasSetName ) != inNames.end() )
+	{
+		return inputSetNamesData;
+	}
+
+	InternedStringVectorDataPtr resultData = inputSetNamesData->copy();
+	resultData->writable().push_back( g_camerasSetName );
+	return resultData;
+}
+
+void LightToCamera::hashSet( const IECore::InternedString &setName, const Gaffer::Context *context, const ScenePlug *parent, IECore::MurmurHash &h ) const
+{
+	if( setName != g_camerasSetName && setName != g_lightsSetName )
+	{
+		h = inPlug()->setPlug()->hash();
+		return;
+	}
+
+	SceneElementProcessor::hashSet( setName, context, parent, h );
+	inPlug()->setPlug()->hash( h );
+
+
+	// This setup is copied from Prune
+
+	// The sets themselves do not depend on the "scene:path"
+	// context entry - the whole point is that they're global.
+	// However, the PathFilter is dependent on scene:path, so
+	// we must remove the path before hashing in the filter in
+	// case we're computed from multiple contexts with different
+	// paths (from a SetFilter for instance). If we didn't do this,
+	// our different hashes would lead to huge numbers of redundant
+	// calls to computeSet() and a huge overhead in recomputing
+	// the same sets repeatedly.
+	//
+	// See further comments in acceptsInput
+	ContextPtr c = filterContext( context );
+	c->remove( ScenePlug::scenePathContextName );
+	Context::Scope s( c.get() );
+	filterPlug()->hash( h );
+}
+
+
+GafferScene::ConstPathMatcherDataPtr LightToCamera::computeSet( const IECore::InternedString &setName, const Gaffer::Context *context, const ScenePlug *parent ) const
+{
+	ConstPathMatcherDataPtr inputSetData = inPlug()->setPlug()->getValue();
+
+	bool isLightSet = setName == g_lightsSetName;
+	bool isCameraSet = setName == g_camerasSetName;
+	if( !( isLightSet || isCameraSet ) )
+	{
+		return inputSetData;
+	}
+
+	ConstPathMatcherDataPtr lightSetData;
+	if( isLightSet )
+	{
+		lightSetData = inputSetData;
+	}
+	else
+	{
+		ContextPtr tmpContext = new Context( *Context::current(), Context::Borrowed );
+		tmpContext->set( ScenePlug::setNameContextName, g_lightsSetName );
+		Context::Scope scopedContext( tmpContext.get() );
+		lightSetData = inPlug()->setPlug()->getValue();
+	}
+
+	const PathMatcher &lightSet = lightSetData->readable();
+
+
+	PathMatcherDataPtr outputSetData = inputSetData->copy();
+	PathMatcher &outputSet = outputSetData->writable();
+
+	ContextPtr tmpContext = filterContext( context );
+	Context::Scope scopedContext( tmpContext.get() );
+
+	for( PathMatcher::RawIterator pIt = lightSet.begin(), peIt = lightSet.end(); pIt != peIt; )
```

Comment:
This logic looks wrong to me, and more complicated than necessary. Here's an alternative stab at it :

```
for( PathMatcher::Iterator pIt = lightSet.begin(), peIt = lightSet.end(); pIt != peIt; ++pIt )
{
    const int m = filterPlug()->getValue();
    if( m & Filter::ExactMatch )
    {
        if( isLightSet )
        {
            outputSet.removePath( *pIt );
        }
        else
        {
            outputSet.addPath( *pIt );
        }
    }
}
```

How does that look?